### PR TITLE
Fix mypy errors (primitives)

### PR DESCRIPTION
**Pull request by Randl**
### Summary

Following discussion, I'm splitting https://github.com/Qiskit/qiskit-terra/pull/8187 by module.

### Details and comments

<details>
<summary> Previous content </summary>
The are ~20 errors left:
```
qiskit/primitives/base_estimator.py:322: error: Invalid index type "Union[int, QuantumCircuit]" for "Tuple[QuantumCircuit, ...]"; expected type "SupportsIndex"  [index]
qiskit/primitives/base_estimator.py:342: error: Invalid index type "Union[int, QuantumCircuit]" for "Tuple[ParameterView, ...]"; expected type "SupportsIndex"  [index]
qiskit/primitives/base_estimator.py:345: error: Invalid index type "Union[int, QuantumCircuit]" for "Tuple[ParameterView, ...]"; expected type "SupportsIndex"  [index]
qiskit/primitives/base_estimator.py:349: error: Invalid index type "Union[int, QuantumCircuit]" for "Tuple[QuantumCircuit, ...]"; expected type "SupportsIndex"  [index]
qiskit/primitives/base_estimator.py:350: error: Invalid index type "Union[int, SparsePauliOp]" for "Tuple[SparsePauliOp, ...]"; expected type "SupportsIndex"  [index]
qiskit/primitives/base_estimator.py:358: error: Value of type variable "SupportsRichComparisonT" of "max" cannot be "Union[int, QuantumCircuit]"  [type-var]
qiskit/primitives/base_estimator.py:358: error: Unsupported operand types for <= ("int" and "QuantumCircuit")  [operator]
qiskit/primitives/base_estimator.py:358: note: Left operand is of type "Union[int, QuantumCircuit]"
qiskit/primitives/base_estimator.py:360: error: Value of type variable "SupportsRichComparisonT" of "max" cannot be "Union[int, QuantumCircuit]"  [type-var]
qiskit/primitives/base_estimator.py:363: error: Value of type variable "SupportsRichComparisonT" of "max" cannot be "Union[int, SparsePauliOp]"  [type-var]
qiskit/primitives/base_estimator.py:363: error: Unsupported operand types for <= ("int" and "SparsePauliOp")  [operator]
qiskit/primitives/base_estimator.py:363: note: Left operand is of type "Union[int, SparsePauliOp]"
qiskit/primitives/base_estimator.py:365: error: Value of type variable "SupportsRichComparisonT" of "max" cannot be "Union[int, SparsePauliOp]"  [type-var]
qiskit/primitives/base_estimator.py:370: error: Argument "circuits" to "_call" of "BaseEstimator" has incompatible type "List[Union[int, QuantumCircuit]]"; expected "Sequence[int]"  [arg-type]
qiskit/primitives/base_estimator.py:371: error: Argument "observables" to "_call" of "BaseEstimator" has incompatible type "List[Union[int, SparsePauliOp]]"; expected "Sequence[int]"  [arg-type]
qiskit/primitives/base_sampler.py:243: error: Invalid index type "Union[int, QuantumCircuit]" for "Tuple[QuantumCircuit, ...]"; expected type "SupportsIndex"  [index]
qiskit/primitives/base_sampler.py:258: error: Invalid index type "Union[int, QuantumCircuit]" for "Tuple[ParameterView, ...]"; expected type "SupportsIndex"  [index]
qiskit/primitives/base_sampler.py:261: error: Invalid index type "Union[int, QuantumCircuit]" for "Tuple[ParameterView, ...]"; expected type "SupportsIndex"  [index]
qiskit/primitives/base_sampler.py:264: error: Value of type variable "SupportsRichComparisonT" of "max" cannot be "Union[int, QuantumCircuit]"  [type-var]
qiskit/primitives/base_sampler.py:264: error: Unsupported operand types for <= ("int" and "QuantumCircuit")  [operator]
qiskit/primitives/base_sampler.py:264: note: Left operand is of type "Union[int, QuantumCircuit]"
qiskit/primitives/base_sampler.py:266: error: Value of type variable "SupportsRichComparisonT" of "max" cannot be "Union[int, QuantumCircuit]"  [type-var]
qiskit/primitives/base_sampler.py:271: error: Argument "circuits" to "_call" of "BaseSampler" has incompatible type "List[Union[int, QuantumCircuit]]"; expected "Sequence[int]"  [arg-type]
```

Large part of those errors can be fixed if the type cast is made explicit (I've added todo in corresponding places).
</details>

Edit (04.10.22): This fixes all remaining `primitives` errors.

### CODE REVIEW

File: qiskit/primitives/base_estimator.py
Comment (rht) [inline qiskit/primitives/base_estimator.py:300]: What does this comment mean? It's ambiguous to me. Either you postpone the type annotation of `circuits`, or you postpone the type casting to `List[int]`.

File: qiskit/primitives/estimator_result.py
Comment (rht) [inline qiskit/primitives/estimator_result.py:44]: `list[dict[str, Any]]` is the better one. The oldest supported Python in qiskit-terra is 3.7 (see https://github.com/Qiskit/qiskit-terra/blob/86fb555871f4811a842380dbf0d54dcda34a4841/setup.py#L81). This means that just using `list` or `dict` works.

**Comment by rht** [comment]
I'm not a qiskit-terra maintainer, so my review does not reflect of the maintainers' preference. But in general, you can use `list`, `dict` directly instead of having to import `List`, `Dict`. As stated, the oldest supported Python version is 3.7, and so it will work fine. This applies to the other Mypy PRs you have.

File: qiskit/primitives/base_estimator.py
Comment (Randl) [reply qiskit/primitives/base_estimator.py:300]: Initially `circuits` is typed as `Sequence[int | QuantumCircuit]`, but after this line it is `List[int]`. This need to be known by `mypy` since afterward precense of `QuantumCircuit` in type leads to type errors

Comment (rht) [reply qiskit/primitives/base_estimator.py:300]: OK I just saw that you wrote
> Large part of those errors can be fixed if the type cast is made explicit (I've added todo in corresponding places).

So, you meant it needs to be casted. But this context of what you want to do with this line, is not known to the people reading just the code.

Comment (rht) [reply qiskit/primitives/base_estimator.py:300]: Yeah, I think you just need to clarify the comment in the way you just described in your reply.

**Comment by rht** [comment]
The content LGTM.

**Comment by 1ucian0** [comment]
Thanks @rht for your review!  
@ElePT , do you mind also having a look?

File: qiskit/primitives/base/base_primitive.py
Comment (ElePT) [reply qiskit/primitives/base/base_primitive.py:18]: (if you apply my other comments, you might not even need the import from `typing`).

**Review by ElePT — COMMENTED** [review]
Hi @Randl, thanks for opening this PR, I have a few general questions plus some more specific comments I have left in the code:

1. I think that making sure we have correct typehints is very valuable, but given that we currently we don't run any mypy checks, I am not convinced about the value added by the constant use of `cast` (which I know, is sometimes the only way to please mypy without making significant changes in the code). My take here would be to add the typehints but not the `cast`s, as they obfuscate the code and make it less readable. 
2.  I see that you fixed the merge conflicts, but you also reverted your changes on `base_sampler.py` and `base_estimator.py`, was this intentional? I think that they should be included for completeness.

**Comment by Randl** [comment]
@ElePT I think that some code I annotated got removed, that's why it appears that some changes are reverted

**Review by ikkoham — CHANGES_REQUESTED** [review]
Great! I have a minor comment. I think `Sequence` is too broad here.

**Comment by ElePT** [comment]
> @ElePT I think that some code I annotated got removed, that's why it appears that some changes are reverted

Sorry, I misread your comment earlier. The code did not get removed, it got **moved**, so those files are still part of the repo (instead of being `primitives.base_estimator.py`, now they are in `primitives.base.base_estimator.py`). Given that you had already looked at them, could you re-apply the changes?

**Review by ElePT — COMMENTED** [review]
Thanks for applying the suggestions!!

> @ElePT I think that some code I annotated got removed, that's why it appears that some changes are reverted

Sorry, I misread your comment earlier. The code did not get removed, it got **moved**, so those files are still part of the repo (instead of being `primitives.base_estimator.py`, now they are in `primitives.base.base_estimator.py`). Given that you had already looked at them, could you re-apply the changes?

**Review by ElePT — APPROVED** [review]
Never mind, I think that I was looking at an outdated version of the PR locally. LGTM.

**Comment by Randl** [comment]
@ikkoham can you approve please?

**Comment by Randl** [comment]
So anything else required from me before it is merged?

**Comment by ElePT** [comment]
@ikkoham any further changes? if not, can you approve the requested changes?

**Review by ikkoham — APPROVED** [review]
Sorry for late. Thank you for achieving my comments. LGTM!

**Pull request merged**
